axelrod/strategies/axelrod_second.py

```diff
@@ -975,3 +975,294 @@ def strategy(self, opponent: Player) -> Action:
+
+        if inc_parity and to_return == D:
+            self.parity_streak[self.parity_bit] += 1
+
+        return to_return
+
+    def detect_random(self, turn):
+        """
+        Calculates a modified Pearson's Chi Squared statistic on self.history,
+        and returns True (is random) if and only if the statistic is less than
+        or equal to 3.
+
+        Pearson's Chi Squared statistic = sum[ (E_i-O_i)^2 / E_i ], where O_i
+        are the observed matrix values, and E_i is calculated as number (of
+        defects) in the row times the number in the column over (total number
+        in the matrix minus 1).
+
+        We say this is modified because it differs from a usual Chi-Squared
+        test in that:
+
+        - It divides by turns minus 2 to get expected, whereas usually we'd
+          divide by matrix total.  Total equals turns minus 1, unless Defect
+          mode has been entered at any point.
+        - Terms where expected counts are less than 1 get excluded.
+        - There's a check at the beginning on the first cell of the matrix.
+        - There's a check at the beginning for the recorded number of defects.
+
+        """
+        denom = turn - 2
+
+        if self.move_history[0, 0] / denom >= 0.8:
+            return False
+        if self.recorded_defects / denom < 0.25 or self.recorded_defects / denom > 0.75:
+            return False
+
+        expected_matrix = np.outer(self.move_history.sum(axis=1), \
+                                    self.move_history.sum(axis=0))
+
+        chi_squared = 0.0
+        for i in range(4):
+            for j in range(2):
+                expct = expected_matrix[i, j] / denom
+                if expct > 1.0:
+                    chi_squared += (expct - self.move_history[i, j]) ** 2 / expct
+
+        if chi_squared > 3:
+            return False
+        return True
+
+    def detect_streak(self, last_move):
+        """
+        Return if and only if the opponent's last twenty moves are defects.
+        """
+
+        if last_move == D:
+            self.defect_streak += 1
+        else:
+            self.defect_streak = 0
+        if self.defect_streak >= 20:
+            return True
+        return False
+
+    def detect_parity_streak(self, last_move):
+        self.parity_bit = 1 - self.parity_bit # Flip bit
+        if last_move == D:
+            self.parity_streak[self.parity_bit] += 1
+        else:
+            self.parity_streak[self.parity_bit] = 0
+        if self.parity_streak[self.parity_bit] >= self.parity_limit:
+            return True
+
+    def strategy(self, opponent: Player) -> Action:
+        turn = len(self.history) + 1
+
+        if turn == 1:
+            return C
+
+        if self.mode == "Defect":
+            if opponent.history[-1] == D:
+                self.exit_defect_meter += 1
+            else:
+                self.exit_defect_meter -= 3
+            if self.exit_defect_meter >= 11:
+                self.mode = "Normal"
```

Comment:
Coverage is failing because we don't have a test case that runs this code block:

```
axelrod/strategies/axelrod_second.py                       436      4    99%   1196-1199
```

(That's the output of coveralls showing that lines 1196-1199 are not hit.)

Reply:
Oh, interesting!  I'll work on it.